Hide widget buttons in foreign user space

chrislessard opened the pull request:
Hey,

This collection of patches disables all the features in the space overview component (and it's child widgets) that should not be available for interaction when visiting another's user's space. Based on the discussion in and surrounding [1], all these features are hidden, rather than disabled.

Here are screenshots of a foreign user space, with each different feature level enabled. Please note that the pipelines widget is left alone, since it is dealt with in [1]. 

The last patch in the series - 'fix(analyze-overview): Hide add to space button in foreign space' - also modifies behaviour introduced in [2].

[1] https://github.com/fabric8-ui/fabric8-ui/pull/2842
[2] https://github.com/fabric8-ui/fabric8-ui/pull/2814

Prod Features
![2018-05-10-172024_1919x1028_scrot](https://user-images.githubusercontent.com/17230733/39894932-773cb758-5476-11e8-9ccf-a8a0cc4ebbee.png)

Beta Features
![2018-05-10-171948_1917x1025_scrot](https://user-images.githubusercontent.com/17230733/39894906-6258014e-5476-11e8-9c63-f5f058ac9a98.png)

Experimental Features
![experimental](https://user-images.githubusercontent.com/17230733/39894585-5aa6898a-5475-11e8-81a3-16cb19053df7.png)

Internal Features
![internal-features](https://user-images.githubusercontent.com/17230733/39894596-6376735e-5475-11e8-8797-8ec3a82fb977.png)

Deals with:
https://github.com/openshiftio/openshift.io/issues/3423
https://github.com/openshiftio/openshift.io/issues/3397
https://github.com/openshiftio/openshift.io/issues/3396
https://github.com/openshiftio/openshift.io/issues/3386

jiekang commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -3,14 +3,18 @@
             </a>
+          </div>
+          <div class="card-pf-heading-details f8-card-heading-details" user-level>
+            <div *ngIf="userOwnsSpace">
```
Using `ngContainer` instead of `div` for all of these ngIf's would be least likely to affect the appearance of the page. Thoughts?

chrislessard replied:
Yes, that's true. As it stands there's no difference for many of the components, but perhaps it would be a good way to future proof.

jiekang commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.spec.ts:
```diff
@@ -89,6 +90,30 @@ describe('AddCodebaseWidgetComponent', () => {
   });
 
+  it('should enable buttons if the user owns the space', function(this: TestingContext) {
+    this.testedDirective.userOwnsSpace = true;
```
Since this is an `@Input` property, to match the actual scenario as closely as possible they should be injected through the host component template rather than being set this way.

jiekang commented:
> In experimental/internal, there don't seem to be any buttons that can be used to "interact" with another users space. However, in Production & Beta the Pipelines widget still has the "+" button in the top right corner, and the "Add to Space" button that can be used to open the app launcher.

For what it's worth the Pipelines widget is being handled in https://github.com/fabric8-ui/fabric8-ui/pull/2842

aptmac commented:
> For what it's worth the Pipelines widget is being handled in #2842 

Ah didn't see it was a separate PR, disregard my above comment then.

chrislessard commented:
@aptmac I actually did work on the pipelines widget [here](https://github.com/fabric8-ui/fabric8-ui/pull/2842), which is why I've left it alone. Good call with all the other pain points, I was meaning to document those. Always more work to be done ..

aptmac commented on src/app/dashboard-widgets/create-work-item-widget/create-work-item-widget.component.ts:
```diff
@@ -22,12 +22,13 @@ class WorkItemFilter {
   contextPath: Observable<string>;
 
   constructor(
-    private context: Contexts,
```
Nice catch

fabric8cd commented:
@chrislessard fabric8/fabric8-ui:SNAPSHOT-PR-2898-2 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2898-fabric8-ui.badger.fabric8.io

joshuawilson commented:
@dlabrecq please review.

fabric8cd commented:
@chrislessard fabric8/fabric8-ui:SNAPSHOT-PR-2898-3 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2898-fabric8-ui.badger.fabric8.io

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -3,14 +3,18 @@
+              <a class="btn btn-link f8-card-heading-btn-link" (click)="addToSpace.emit()">
+                <i id="user-level-add-to-space-btn" class="pficon pficon-add-circle-o"></i>
+              </a>
+            </ng-container>
```
For a component like work-item-widget.component, ng-container looks like a good fit because it's used in just one place. However, add-codebase-widget.component uses several ng-container tags throughout the HTML layout.

To complicate things, the HTML layout is already wrapped in multiple places for the AppLauncher feature toggle. That forces us to wrap each user-level and default-level directive with multiple ng-container tags.

Considering most of the functionality in this component is being wrapped with ng-container, could we create an alternate, simplified layout for this scenario? I believe that would help make the code more readable and maintainable.

Alternatively, we could remove the AppLuncher feature -- looks like that's not used anymore?

dlabrecq commented on src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.html:
```diff
@@ -2,65 +2,70 @@
-                  (click)="saveDescription()" >
-            Save
-          </button>
+    <ng-container *ngIf="userOwnsSpace; else uneditableDescription">
```
Most of the functionality for this component has been wrapped with ng-container. This could be another candidate for an alternate, more simplified layout.

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -3,14 +3,18 @@
+          </ng-container>
+          <div class="card-pf-heading-details f8-card-heading-details" user-level>
+            <ng-container *ngIf="userOwnsSpace">
+              <a class="btn btn-link f8-card-heading-btn-link" (click)="addToSpace.emit()">
```
Other than the ID, this button is exactly the same as line 7. I see that the AppLauncher feature has been changed, which makes me question if this should be removed?

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -20,16 +24,20 @@ <h2 class="card-pf-title">
+          <h3>This Space has no Codebases</h3>
+          <div class="f8-blank-slate-main-action">
+            <ng-container *ngIf="userOwnsSpace">
+              <button id="user-level-spacehome-codebases-create-button" class="btn btn-primary btn-lg" (click)="addToSpace.emit()">Add Codebase</button>
```
Other than the ID, this button is exactly the same as line 30. I see that the AppLauncher feature has been changed, which makes me question if this should be removed?

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -56,17 +65,21 @@ <h3>This Space has no Codebases</h3>
+        </div>
+        <div class="card-pf-heading-details f8-card-heading-details" user-level>
+          <ng-container *ngIf="userOwnsSpace">
+            <a class="btn btn-link f8-card-heading-btn-link" (click)="addToSpace.emit()">
```
Same functionality as line 70

christianvogt commented on src/app/space/analyze/analyze-overview/analyze-overview.component.html:
```diff
@@ -2,24 +2,28 @@
+          <div *ngIf="userOwnsSpace()">
+            <button *ngIf="this.authentication.isLoggedIn()" id="default-level-analyze-overview-dashboard-create-space-button" class="btn btn-primary btn-lg f8-dashboard-masthead-create-application pull-right"
+              (click)="openForgeWizard(addSpace)" default-level> Create an Application </button>
+            <button *ngIf="this.authentication.isLoggedIn()" id="user-level-analyze-overview-dashboard-create-space-button" class="btn btn-primary btn-lg f8-dashboard-masthead-create-application pull-right"
```
We should avoid assigning IDs. Especially when only used for testing purposes. Consider using a custom attribute such as `data-tid` (tid = test id).  You can write a selector such as `*[data-tid=foo]`. We could take this one step further and probably even strip these attributes in the future from production builds.

chrislessard replied:
Would you mind elaborating a bit more on why its bad to test that way ?

christianvogt replied:
There's nothing wrong with doing it this way and I won't disapprove the change as it stands now. When using IDs we must make them globally unique (Which you have done). But for a test id in a custom attribute we only need them unique within the component. We should establish a pattern to follow when creating test IDs for unit or automated tests.
I wasn't sure if it had ever been discussed for osio or not.

I'll bring it up separately as another tech debt item with the team.

joshuawilson replied:
ID's should only be used for test references, at least as much as possible. There are a few instances where you need it for something like a form label. A Test ID is fine for tests too, though we haven't done it as we don't use ID's for references in the code.

fabric8cd commented:
@chrislessard fabric8/fabric8-ui:SNAPSHOT-PR-2898-4 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2898-fabric8-ui.badger.fabric8.io

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -3,14 +3,18 @@
-                <i class="pficon pficon-add-circle-o"></i>
+              <i id="default-level-add-to-space-btn" class="pficon pficon-add-circle-o"></i>
             </a>
+          </div>
```
Lines 5-18.

The user-level directive shouldn't be located inside the default-level directive.

Other than the link ID, the code for this feature toggle appears to be the same. Can we just remove it?

```
<div class="card-pf-heading-details f8-card-heading-details" default-level>
          <div *ngIf="userOwnsSpace">
            <a class="btn btn-link f8-card-heading-btn-link" (click)="addToSpace.emit()">
              <i id="default-level-add-to-space-btn" class="pficon pficon-add-circle-o"></i>
            </a>
          </div>
          <div class="card-pf-heading-details f8-card-heading-details" user-level>
            <div *ngIf="userOwnsSpace">
              <a class="btn btn-link f8-card-heading-btn-link" (click)="addToSpace.emit()">
                <i id="user-level-add-to-space-btn" class="pficon pficon-add-circle-o"></i>
              </a>
            </div>
          </div>
</div>
```

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -20,16 +24,20 @@ <h2 class="card-pf-title">
+              <button id="user-level-spacehome-codebases-create-button" class="btn btn-primary btn-lg" (click)="addToSpace.emit()">Add Codebase</button>
             </div>
+          </div>
         </div>
```
Lines 25-41

Again, the only difference here is the button ID. We should clean this up and just remove the AppLauncher feature.

```
<div class="f8-blank-slate-card" *ngIf="codebases.length === 0" default-level>
          <h3>This Space has no Codebases</h3>
          <div class="f8-blank-slate-main-action">
            <div *ngIf="userOwnsSpace">
              <button id="default-level-spacehome-codebases-create-button" class="btn btn-primary btn-lg" (click)="addToSpace.emit()">Add Codebase</button>
            </div>
          </div>
        </div>
        <div class="f8-blank-slate-card" *ngIf="codebases.length === 0" user-level>
          <h3>This Space has no Codebases</h3>
          <div class="f8-blank-slate-main-action">
            <div *ngIf="userOwnsSpace">
              <button id="user-level-spacehome-codebases-create-button" class="btn btn-primary btn-lg" (click)="addToSpace.emit()">Add Codebase</button>
            </div>
          </div>
</div>
```

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -56,17 +65,22 @@ <h3>This Space has no Codebases</h3>
+              <i class="pficon pficon-add-circle-o"></i>
+            </a>
+          </div>
+        </div>
```
Lines 68-82

The only difference is the link ID.

```
<div class="card-pf-heading-details f8-card-heading-details" default-level>
          <div *ngIf="userOwnsSpace">

            <a id="spacehome-codebases-add-button" class="btn btn-link f8-card-heading-btn-link" (click)="addToSpace.emit()">
              <i class="pficon pficon-add-circle-o"></i>
            </a>
          </div>
        </div>
        <div class="card-pf-heading-details f8-card-heading-details" user-level>
          <div *ngIf="userOwnsSpace">
            <a class="btn btn-link f8-card-heading-btn-link" (click)="addToSpace.emit()">
              <i class="pficon pficon-add-circle-o"></i>
            </a>
          </div>
</div>
```

dlabrecq commented on src/app/dashboard-widgets/add-codebase-widget/add-codebase-widget.component.html:
```diff
@@ -77,13 +91,17 @@ <h2 class="card-pf-title">
+              <button id="user-level-spacehome-my-codebases-create-button" class="btn btn-primary btn-lg" (click)="addToSpace.emit()">Add Codebase</button>
+            </div>
           </div>
         </div>
```
Lines 91-106

The only change is the button ID. Lets remove the AppLauncher feature here as well.

```
<div class="f8-blank-slate-card" *ngIf="codebases.length === 0" default-level>
          <h3>This Space has no Codebases</h3>
          <div class="f8-blank-slate-main-action">
            <div *ngIf="userOwnsSpace">
              <button id="default-level-spacehome-my-codebases-create-button" class="btn btn-primary btn-lg" (click)="addToSpace.emit()">Add Codebase</button>
            </div>
          </div>
        </div>
        <div class="f8-blank-slate-card" *ngIf="codebases.length === 0" user-level>
          <h3>This Space has no Codebases</h3>
          <div class="f8-blank-slate-main-action">
            <div *ngIf="userOwnsSpace">
              <button id="user-level-spacehome-my-codebases-create-button" class="btn btn-primary btn-lg" (click)="addToSpace.emit()">Add Codebase</button>
            </div>
          </div>
</div>
```

dlabrecq commented on src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.html:
```diff
@@ -3,16 +3,19 @@
   <div class="col-sm-9 f8-space-masthead-description">
     <!-- There must be no spaces here, otherwise we end up with random whitespace in the editable-->
-    <div class="form-group">
+    <div *ngIf="userOwnsSpace; else userDoesNotOwnSpaceDescription" class="form-group">
```
This looks much cleaner 👍

dlabrecq commented on src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.html:
```diff
@@ -3,16 +3,19 @@
+    </ng-template>
   </div>
-  <div class="f8-space-masthead-vertical-divider"></div>
+  <div *ngIf="userOwnsSpace" class="f8-space-masthead-vertical-divider"></div>
```
I believe we should keep the divider considering we're still displaying the description -- just not allowing editing.

chrislessard commented:
@dlabrecq 

> "I believe we should keep the divider considering we're still displaying the description -- just not allowing editing."

Good catch, I changed the if condition on this one. 

As for the comments about the AppLauncher, the purpose of this PR is strictly to hide buttons and functionality that shouldn't be available when in another user's space. Why I agree that the code could be simplified, I belief this is work for a separate PR. What do you think ?

dlabrecq commented:
I believe we should clean this up and remove the AppLauncher feature flag here because we're introducing new conditionals to hide duplicated behavior.

fabric8cd commented:
@chrislessard fabric8/fabric8-ui:SNAPSHOT-PR-2898-5 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2898-fabric8-ui.badger.fabric8.io

jiekang commented:
I'd prefer all of the changes for the 'add codebase' widget to be done separately. Two PRs, one to clean up the feature flag related code and one to disable the button, or a combined PR are both fine. The changes for the rest of the widgets/components look okay.

chrislessard commented:
@dlabrecq @joshuawilson @jiekang Following Jie's advice I've dropped all changes to the add-codebase-widget from this PR. I've moved that work over to a new branch, which will address Dan's comments. Dan, I'll request you as a reviewer once I've opened it.

fabric8cd commented:
@chrislessard fabric8/fabric8-ui:SNAPSHOT-PR-2898-8 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2898-fabric8-ui.badger.fabric8.io
